### ISSUE TITLE
 Finetuning HSMR model

### ISSUE THREAD

**Issue by mohamedyousuf1** [IsshikiHugh/HSMR#19]
At first, thank you for sharing your valuable work.
I am trying to fine-tune the model, but I am facing difficulty in annotating the dataset. Based on studying the trained data, we have two needed inputs (a. Segmenting mask  b. The following keys for the betas, pose, and camera parameters):
`['keypoints_2d', 'keypoints_3d', 'center', 'scale', 'personid', 'extra_info', 'orig_poses', 'orig_betas', 'orig_trans', 'flip_poses', 'flip_betas', 'flip_trans', 'orig_pve_max', 'flip_pve_max', 'orig_pve_mean', 'flip_pve_mean']`

Here is a table that shows my understanding of how to obtain each key:

| Key                 | Description                            | How It Was Obtained (in HSMR)                                                                            |
| ------------------- | -------------------------------------- | -------------------------------------------------------------------------------------------------------- |
| `keypoints_2d`      | 2D joints used for reprojection loss   | Projected from SKEL mesh using camera intrinsics; refined via SKELify optimization (fit to 2D keypoints) |
| `keypoints_3d`      | 3D joint positions                     | Extracted from SKEL mesh joint locations                                                                 |
| `center`            | Image-space center of the person       | Computed from projected 2D SKEL joints' bounding box center                                              |
| `scale`             | Person scale (used for cropping)       | Computed from SKEL joint-based 2D bounding box size                                                      |
| `personid`          | Unique person ID                       | Derived from sample indexing of the original datasets                                                    |
| `extra_info`        | Metadata (e.g., source dataset)        | Retained from the metadata of the original datasets used (e.g., COCO, H36M)                              |
| `orig_poses`        | SKEL pose parameters (rotations)       | Obtained from SMPL → SKEL conversion using offline batch optimization (SKELify-style)                    |
| `orig_betas`        | SKEL shape parameters                  | Also from offline SMPL → SKEL conversion                                                                 |
| `orig_trans`        | Global translation vector              | Obtained alongside pose and shape from the SKEL conversion process                                       |
| `flip_poses`        | Pose for horizontally flipped image    | SKELify applied independently to mirrored image; not derived by param flip                               |
| `flip_betas`        | Shape from flipped image               | Independently estimated using SKELify on mirrored image                                                  |
| `flip_trans`        | Translation from flipped image         | Also from SKELify on mirrored image                                                                      |
| `orig_pve_max`      | Max per-vertex error (quality measure) | Max Euclidean distance between SMPL and SKEL mesh vertices                                               |
| `flip_pve_max`      | Max PVE for flipped version            | Same as above, for flipped version                                                                       |
| `orig_pve_mean`     | Mean per-vertex error                  | Mean Euclidean distance between SMPL and SKEL mesh vertices                                              |
| `flip_pve_mean`     | Mean PVE for flipped case              | Same as above, for flipped version                                                                       |
| `segmentation_mask` | Person silhouette                      | Derived using segmentation tools from original datasets or pretrained models                             |

I understood from your Supplementary Material that the flip keys is issued separately, center and scale come from the detection. Now, how can I obtain keypoints, original poses, betas, and trans for unannotated images since I don't see how to make the conversion from SMPL → SKEL, and SKELify? It would be great if you could provide the steps on how to fine-tune using custom images , with annotating the data to match the preprocessing keys.

Thank you!

**Comment by IsshikiHugh** [comment]
Hi, thank you for being interested in HSMR!

First, I had to say I don't know all the details here either. The data are directly modified from HMR2.0's data. For example, I don't know how `segmentation_mask` helps. But this won't influence HSMR training.


And about your question, if I understand correctly, you want to create a new dataset without annotation accurate SKEL parameters 3D annotation, but you might have SMPL parameters , right?

In that case:

1.  `keypoints_2d ` is not be obtained from the SKEL parameters. And SKELify use 2D keypoints to refine 3D annotations, just the opposite of what you said. You might want to obtain them from 1) accurate SMPL reprojection; 2) the original 2D keypoints annotation (you might not have all of them, in that case you need to map and mask the valid joints); 3) using keypoints detection tools to annotate them.
2. For bounding box (`center` and `scale`), it's better to calculate them by reprojecting the mesh vertices, instead of only joints. It's easy to understand, joints are usually around the geometry center of the mesh, the bounding boxes will be "too tight" if you calculate them according to the joints2d.
3. SKELify will not be used to initialize the data. They are used during training in our setting. Since SKELify is a fully 2D optimization, using SKELify to build the data will cause troubles. The 3D optimization is something similar to [this](https://github.com/MarilynKeller/SKEL/blob/master/examples/align_to_SMPL_frame.py) , but not exactly the same. The main optimization settings we used are [here](https://github.com/IsshikiHugh/HSMR/tree/main/lib/body_models/skel/alignment), but I didn't upload the entrance file. Since you request this, I will find a time slot to release the related entrance code. But you can also try to adapt them yourselves.

I had to say I never tried to create a training data from only image for HSMR.
If you have 3D annotation like SMPL parameters, then it should be something like a SMPL-image dataset. Otherwise, you might need extra tools to get the information needed. These kinds of data can be more unreliable than what we had for training HSMR.

Hope this helps!

**Comment by IsshikiHugh** [comment]
> Thank you for your comment. Currently, I only have images and no annotations; I plan to make the annotations manually or semi-automatically. To my understanding, I need to understand how HMR 2.0 annotated its data and then feed that into your model. Is my understanding correct?

Basically, yes. But you might needn't to first get SMPL annotations and then convert them to SKEL. But for other annotations, you should be right.

**Comment by IsshikiHugh** [comment]
Hi, it seems like I don't have much time to do the code review these days.
But I'd like to share the original entrance code that I had used.

Since you might don't have the same format of the inputs, so I didn't upload the inputs format process functions.
Some of the code there (like some `lock` things) are for supporting optimization on different GPUs.

I only did simple cleaning, so I can't ensure the code is clear enough or suitable for general purpose.

But I think this can serve as a detailed reference for you.


```python
import time
import trimesh
import numpy as np
import smplx
import torch

from pathlib import Path

from lib.utils.media import *
from lib.utils.vis import Wis3D
from lib.utils.data import disassemble_dict, assemble_dict
from lib.platform import PM
from lib.kits.debug import *
from lib.body_models.skel.alignment.aligner import SkelFitter
from lib.info.show import show_distribution

DEVICES = ['cuda:0']
ANALYZE_DEVICES = ['cuda:0']
GENDER = 'male'


def fit_skel(smpl_params, device=DEVICES[0], batch_size=1, skip_first_n_round=0, output_fn=None, variant_exp=''):
    ''' Use the SMPL parameters to fit the skeleton. '''
    # Prepare the input data.
    L_ = smpl_params['betas'].shape[0]
    size_round = batch_size
    n_round = (L_-1)// size_round + 1
    skel_params_list = []

    print(f'Totally {n_round} rounds.')
    print(f'Will skip {skip_first_n_round} rounds')
    for i in tqdm(range(n_round)):
        output_fn_i = output_fn.parent / 'intermediate' / f'{output_fn.stem}_{i}|{n_round}.pt'
        output_fn_i.parent.mkdir(parents=True, exist_ok=True)
        output_fn_i_tmp = output_fn.parent / 'intermediate' / f'{output_fn.stem}_{i}|{n_round}.lock'
        if i < skip_first_n_round or output_fn_i.exists() or output_fn_i_tmp.exists():
            if not output_fn_i.exists():
                get_logger().warning(f'Skip round {i} because {output_fn_i} does not exist.')
                continue

            skel_params_ = torch.load(output_fn_i)

            if np.isnan(skel_params_['poses']).any() \
            or np.isnan(skel_params_['betas']).any() \
            or np.isnan(skel_params_['trans']).any():
                output_fn_i.unlink()
                get_logger().warning(f'Rerun round {i} because {output_fn_i} contains NaN.')
                # set_trace()
            else:
                skel_params_list.append(skel_params_)
                continue
        else:
            pass

        output_fn_i_tmp.touch()  # Lock the file.
        sid = i * size_round
        eid = min((i+1) * size_round, L_)
        L = eid - sid
        while(True):
            betas_in = torch.from_numpy(smpl_params['betas'][sid:eid].reshape(L, 10)).to(device)
            poses_in = torch.from_numpy(np.concatenate([smpl_params['global_orient'][sid:eid], smpl_params['body_pose'][sid:eid]], axis=1).reshape(L, 24*3)).to(device)
            nan_masks = torch.isnan(betas_in).any(dim=1) | torch.isnan(poses_in).any(dim=1)
            betas_in[nan_masks, :] = torch.nan  # uniformize the nan values
            poses_in[nan_masks, :] = torch.nan  # uniformize the nan values
            valid_masks = ~nan_masks
            valid_betas = betas_in[valid_masks]
            valid_poses = poses_in[valid_masks]
            valid_L = valid_betas.shape[0]
            valid_masks = to_numpy(valid_masks)

            try:
                output_fn_i_tmp.touch()  # Lock the file.
                params = {
                    'trans_in'        : torch.zeros((valid_L, 3)).to(device),
                    'betas_in'        : valid_betas,
                    'poses_in'        : valid_poses,
                    'batch_size'      : batch_size,
                    'skel_data_init'  : None,
                    'force_recompute' : True,
                    'debug'           : False,  # Only run the fit on the first minibach to test if `True`.
                    'watch_frame'     : 0,
                    'opt_sequence'    : False,
                }

                # Fit the skeleton.
                skel_fitter = SkelFitter(GENDER, device=device)
                skel_params_valid = skel_fitter.run_fit(**params, variant_exp=variant_exp)

                # skel_params_valid to skel_params_, i.e. embed the valid results to the full results.
                skel_params_ = {
                        'poses': np.full((L, 46), 0).astype(np.float32),
                        'betas': np.full((L, 10), 0).astype(np.float32),
                        'trans': np.full((L, 3), 0).astype(np.float32),
                    }
                skel_params_['poses'][valid_masks] = skel_params_valid['poses']
                skel_params_['betas'][valid_masks] = skel_params_valid['betas']
                skel_params_['trans'][valid_masks] = skel_params_valid['trans']

                break
            except Exception as e:
                print(f'Error: {e}')
            finally:
                output_fn_i_tmp.unlink()  # Unlock the file.

        skel_params_list.append(skel_params_)
        # Save the intermediate results.
        torch.save(skel_params_, output_fn_i)

    skel_params = assemble_dict(skel_params_list, expand_dim=False, keys=['poses', 'betas', 'trans'])
    return skel_params


def fit_skel_batched(smpl_params, batch_size, output_fn, skip_first_n_round=0, variant_exp=''):
    import time

    t_s = time.time()
    # smpl_params_list = disassemble_dict(smpl_params, keep_dim=True)
    # smpl_params = assemble_dict(smpl_params_list, expand_dim=False, keys=['global_orient', 'body_pose', 'betas'])
    results = fit_skel(
            smpl_params,
            device             = DEVICES[0],
            batch_size         = batch_size,
            skip_first_n_round = skip_first_n_round,
            output_fn          = output_fn,
            variant_exp        = variant_exp,
        )
    t_e = time.time()
    L = smpl_params['betas'].shape[0]
    print(f'=====================')
    print(f'⏰ Batched ({batch_size}) fitting takes {t_e - t_s:.2f} seconds for {L} frames.')
    print(f'=====================')
    return results
```

**Comment by mohamedyousuf1** [comment]
Thank you for your comment and your help!
I would close this one now, if something comes up, I will let you know.
